### PR TITLE
FOGL-995/996 - South plugins reconfig revamped

### PR DESCRIPTION
**Pull request by amarendra-dianomic**
If plugin reconfigure requires a restart, we have to first stop it. Therefore, this code block duplicates the plugin_shutdown() almost entirely, except for logging. Therefore, I have added a TODO item to decide, may be later, if we want to call plugin_shutdown() here and avoid code duplication.

### CODE REVIEW

File: python/foglamp/plugins/south/cc2650async/cc2650async.py [hunk 1]
```diff
@@ -336,9 +336,43 @@ def plugin_reconfigure(handle, new_config):
...
 
+            tag.disconnect()
+            _LOGGER.info('SensorTagCC2650 (async) {} Disconnected.'.format(bluetooth_adr))
+
+        new_handle = plugin_init(new_config)
+        new_handle['restart'] = 'yes'
+        _LOGGER.info("Restarting CC2650ASYN plugin due to change in configuration keys [{}]".format(', '.join(diff)))
+    else:
+        new_handle = copy.deepcopy(handle)
+        new_handle['restart'] = 'no'
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/cc2650async/cc2650async.py]: better to set True / False ? as its not saved to config (which accepts literal) but run time decision

File: python/foglamp/plugins/south/cc2650poll/cc2650poll.py [hunk 2]
```diff
@@ -259,10 +257,31 @@ def plugin_reconfigure(handle, new_config):
...
 
-    new_handle = plugin_init(new_config)
+    # Find diff between old config and new config
+    diff = list()
+    for key in new_config:
+        if key in handle:
+            if handle[key] != new_config[key]:
+                diff.append(key)
+        else:
+            diff.append(key)
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/cc2650poll/cc2650poll.py]: as above

File: python/foglamp/plugins/south/cc2650poll/cc2650poll.py [hunk 3]
```diff
@@ -259,10 +257,31 @@ def plugin_reconfigure(handle, new_config):
...
+
+    # Plugin should re-initialize and restart if key configuration is changed
+    if 'bluetoothAddress' in diff:
+        # TODO: Investigate if a common stop_plugin() method, shared by plugin_shutdown(), required.
+        if 'tag' in handle:
+            bluetooth_adr = handle['bluetoothAddress']['value']
+            tag = handle['tag']
+            tag.disconnect()
+            _LOGGER.info('SensorTagCC2650 {} Disconnected.'.format(bluetooth_adr))
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/cc2650poll/cc2650poll.py]: as above! ALso, add a newline break

File: python/foglamp/plugins/south/coap_listen/coap_listen.py [hunk 4]
```diff
@@ -115,8 +115,31 @@ def plugin_reconfigure(handle, new_config):
...
+    _LOGGER.info("Old config for COAP plugin {} \n new config {}".format(handle, new_config))
+
+    # Find diff between old config and new config
+    diff = list()
+    for key in new_config:
+        if key in handle:
+            if handle[key] != new_config[key]:
+                diff.append(key)
+        else:
+            diff.append(key)
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/coap_listen/coap_listen.py]: as above

File: python/foglamp/plugins/south/coap_listen/coap_listen.py [hunk 5]
```diff
@@ -115,8 +115,31 @@ def plugin_reconfigure(handle, new_config):
...
+
+    # Plugin should re-initialize and restart if key configuration is changed
+    if 'port' in diff or 'uri' in diff:
+        # TODO: Investigate if a common stop_plugin() method, shared by plugin_shutdown(), required.
+        try:
+            asyncio.ensure_future(aiocoap_ctx.shutdown())
+        except Exception as ex:
+            _LOGGER.exception('Error in shutting down COAP plugin {}'.format(str(ex)))
+            raise
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/coap_listen/coap_listen.py]: above block in if is copy of plugin_shutdown?

File: python/foglamp/plugins/south/http_south/http_south.py [hunk 6]
```diff
@@ -107,10 +109,39 @@ def plugin_reconfigure(handle, new_config):
...
 
-    new_handle = plugin_init(new_config)
+    # Find diff between old config and new config
+    diff = list()
+    for key in new_config:
+        if key in handle:
+            if handle[key] != new_config[key]:
+                diff.append(key)
+        else:
+            diff.append(key)
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/http_south/http_south.py]: as above

File: python/foglamp/plugins/south/http_south/http_south.py [hunk 7]
```diff
@@ -107,10 +109,39 @@ def plugin_reconfigure(handle, new_config):
...
+            server = handle['server']
+
+            server.close()
+            asyncio.ensure_future(server.wait_closed())
+            asyncio.ensure_future(app.shutdown())
+            asyncio.ensure_future(handler.shutdown(60.0))
+            asyncio.ensure_future(app.cleanup())
+        except Exception as e:
+            _LOGGER.exception(str(e))
+            raise
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/http_south/http_south.py]: copy paste of plugin_shutdown?

File: python/foglamp/plugins/south/http_south/http_south.py [hunk 8]
```diff
@@ -107,10 +109,39 @@ def plugin_reconfigure(handle, new_config):
...
+        new_handle = plugin_init(new_config)
+        new_handle['restart'] = 'yes'
+        _LOGGER.info("Restarting HTTP_SOUTH plugin due to change in configuration keys [{}]".format(', '.join(diff)))
+    else:
+        new_handle = copy.deepcopy(handle)
+        new_handle['restart'] = 'no'
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/http_south/http_south.py]: boolean

File: python/foglamp/plugins/south/http_south/http_south.py [hunk 9]
```diff
@@ -107,10 +109,39 @@ def plugin_reconfigure(handle, new_config):
...
+            server.close()
+            asyncio.ensure_future(server.wait_closed())
+            asyncio.ensure_future(app.shutdown())
+            asyncio.ensure_future(handler.shutdown(60.0))
+            asyncio.ensure_future(app.cleanup())
+        except Exception as e:
+            _LOGGER.exception(str(e))
+            raise
+        new_handle = plugin_init(new_config)
+        new_handle['restart'] = 'yes'
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/http_south/http_south.py]: boolean

File: python/foglamp/plugins/south/http_south/http_south.py [hunk 10]
```diff
@@ -124,10 +155,10 @@ def plugin_shutdown(data):
         asyncio.ensure_future(app.shutdown())
         asyncio.ensure_future(handler.shutdown(60.0))
         asyncio.ensure_future(app.cleanup())
-        _LOGGER.info('South HTTP plugin shut down.')
     except Exception as e:
         _LOGGER.exception(str(e))
         raise
+    _LOGGER.info('South HTTP plugin shut down.')
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/http_south/http_south.py]: will this make a diff? it was last line in try block

File: python/foglamp/plugins/south/http_south/http_south.py [hunk 11]
```diff
@@ -148,16 +179,11 @@ class HttpSouthIngest(object):
                         "timestamp": "2017-01-02T01:02:03.23232Z-05:00",
                         "asset": "pump1",
                         "key": "80a43623-ebe5-40d6-8d80-3f892da9b3b4",
-                        "readings": {
-                            "velocity": "500",
-                            "temperature": {
-                                "value": "32",
-                                "unit": "kelvin"
-                            }
+                        "readings": {"humidity": 0.0, "temperature": -40.0}
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/http_south/http_south.py]: was there a problem in old docstring sample?

File: python/foglamp/services/south/ingest.py [hunk 12]
```diff
@@ -516,7 +516,9 @@ def is_available(cls) -> bool:
                 An invalid value was provided
         """
         if cls._stop:
-            raise RuntimeError('The South server is stopping')
+            _LOGGER.warning('The South server is stopping')
+            return
+            # raise RuntimeError('The South server is stopping')
```
Comment (praveen-garg) [inline python/foglamp/services/south/ingest.py]: remove comment

File: python/foglamp/services/south/server.py [hunk 13]
```diff
@@ -8,6 +8,9 @@
 
 import asyncio
 import signal
+
+import pexpect
```
Comment (praveen-garg) [inline python/foglamp/services/south/server.py]: why do we need pexpect here?

File: python/foglamp/services/south/server.py [hunk 14]
```diff
@@ -61,6 +64,8 @@ class Server(FoglampMicroservice):
     """The value that is returned by the plugin_init"""
 
     _type = "Southbound"
+    _task_async = None
+    _task_poll = None
```
Comment (praveen-garg) [inline python/foglamp/services/south/server.py]: we already have a mode, can not we have _task only?

File: python/foglamp/services/south/server.py [hunk 15]
```diff
@@ -117,29 +122,34 @@ class Server(FoglampMicroservice):
...
             elif self._plugin_info['mode'] == 'poll':
-                asyncio.ensure_future(self._exec_plugin_poll(config))
-
-        except (Exception, KeyError) as ex:
+                self._task_poll = asyncio.ensure_future(self._exec_plugin_poll())
+        except asyncio.CancelledError:
+            pass
+        except exceptions.DataRetrievalError:
+            _LOGGER.exception('Data retreival error in plugin {}'.format(self._name))
+        except (Exception, KeyError, pexpect.exceptions.TIMEOUT) as ex:
```
Comment (praveen-garg) [inline python/foglamp/services/south/server.py]: better to capture pexpect.exceptions.TIMEOUT at plugin level in timeouterror and raise that with info

File: python/foglamp/services/south/server.py [hunk 16]
```diff
@@ -117,29 +122,34 @@ class Server(FoglampMicroservice):
...
             _LOGGER.exception(error)
             print(error, str(ex))
             asyncio.ensure_future(self._stop(loop))
 
-    async def _exec_plugin_async(self, config) -> None:
+    async def _exec_plugin_async(self) -> None:
         """Executes async type plugin
         """
-        await Ingest.start(self._core_management_host, self._core_management_port)
+        # await Ingest.start(self._core_management_host, self._core_management_port)
```
Comment (praveen-garg) [inline python/foglamp/services/south/server.py]: remove comment

File: python/foglamp/services/south/server.py [hunk 17]
```diff
@@ -117,29 +122,34 @@ class Server(FoglampMicroservice):
...
         self._plugin.plugin_start(self._plugin_handle)
 
-    async def _exec_plugin_poll(self, config) -> None:
+    async def _exec_plugin_poll(self) -> None:
         """Executes poll type plugin
         """
-        await Ingest.start(self._core_management_host, self._core_management_port)
+        # await Ingest.start(self._core_management_host, self._core_management_port)
```
Comment (praveen-garg) [inline python/foglamp/services/south/server.py]: remove comment

File: python/foglamp/services/south/server.py [hunk 18]
```diff
@@ -158,13 +168,13 @@ class Server(FoglampMicroservice):
...
                 # pollInterval is expressed in milliseconds
-                sleep_seconds = int(config['pollInterval']['value']) / 1000.0
+                sleep_seconds = int(self._plugin_handle['pollInterval']['value']) / 1000.0
                 await asyncio.sleep(sleep_seconds)
                 # If successful, then set retry count back to 1, meaning that only in case of 3 successive failures, exit.
                 try_count = 1
             except KeyError as ex:
                 _LOGGER.exception('Keyerror plugin {} : {}'.format(self._name, str(ex)))
-            except (Exception, RuntimeError) as ex:
+            except (Exception, RuntimeError, exceptions.DataRetrievalError) as ex:
```
Comment (praveen-garg) [inline python/foglamp/services/south/server.py]: plugin specific? ` exceptions.DataRetrievalError` if Yes, we should capture in pythonic way to keep things independent

File: python/foglamp/services/south/server.py [hunk 19]
```diff
@@ -205,10 +205,18 @@ def run(self):
...
+            if self._task_async:
+                self._task_async.cancel()
+            if self._task_poll:
+                self._task_poll.cancel()
+            # Cancel all pending asyncio tasks after a timeout occurs
+            done, pending = await asyncio.wait(asyncio.Task.all_tasks(), timeout=5)
+            for task_pending in pending:
+                task_pending.cancel()
+            await asyncio.sleep(2)
+        except (asyncio.CancelledError, exceptions.DataRetrievalError, pexpect.exceptions.TIMEOUT):
```
Comment (praveen-garg) [inline python/foglamp/services/south/server.py]: as above!

File: python/foglamp/services/south/server.py [hunk 20]
```diff
@@ -205,10 +205,18 @@ def run(self):
...
-        # Cancel all pending asyncio tasks after a timeout occurs
-        done, pending = await asyncio.wait(asyncio.Task.all_tasks(), timeout=5)
-        for task_pending in pending:
-            task_pending.cancel()
+        try:
+            if self._task_async:
+                self._task_async.cancel()
+            if self._task_poll:
+                self._task_poll.cancel()
+            # Cancel all pending asyncio tasks after a timeout occurs
```
Comment (praveen-garg) [inline python/foglamp/services/south/server.py]: if we will have 1 var, `self._task.cancel()`

File: python/foglamp/services/south/server.py [hunk 21]
```diff
@@ -234,36 +242,28 @@ def run(self):
...
+                if self._plugin_info['mode'] == 'async':
+                    self._task_async.cancel()
+                    self._task_async = asyncio.ensure_future(self._exec_plugin_async())
+                elif self._plugin_info['mode'] == 'poll':
+                    self._task_poll.cancel()
+                    self._task_poll = asyncio.ensure_future(self._exec_plugin_poll())
+                await asyncio.sleep(2)
+        except asyncio.CancelledError:
+            pass
+        except (exceptions.DataRetrievalError, pexpect.exceptions.TIMEOUT):
```
Comment (praveen-garg) [inline python/foglamp/services/south/server.py]: as above

File: python/foglamp/plugins/south/cc2650async/cc2650async.py (hunk 1 above)
Comment (amarendra-dianomic) [reply python/foglamp/plugins/south/cc2650async/cc2650async.py]: For better self documentation purpose, yes/no has been chosen over True/False.

File: python/foglamp/plugins/south/coap_listen/coap_listen.py (hunk 5 above)
Comment (amarendra-dianomic) [reply python/foglamp/plugins/south/coap_listen/coap_listen.py]: yes

File: python/foglamp/plugins/south/cc2650async/cc2650async.py [hunk 22]
```diff
@@ -336,9 +336,43 @@ def plugin_reconfigure(handle, new_config):
...
+            tag.char_write_cmd(handle['characteristics']['humidity']['configuration']['handle'], char_disable)
+            tag.char_write_cmd(handle['characteristics']['pressure']['configuration']['handle'], char_disable)
+            tag.char_write_cmd(handle['characteristics']['movement']['configuration']['handle'], movement_disable)
+
+            # Disable notification
+            for notification_handle in handle['notification_handles']:
+                tag.char_write_cmd(notification_handle, notification_disable)
 
+            tag.disconnect()
+            _LOGGER.info('SensorTagCC2650 (async) {} Disconnected.'.format(bluetooth_adr))
```
[the remark this answers is not in the document]
Comment (amarendra-dianomic) [reply python/foglamp/plugins/south/cc2650async/cc2650async.py]: yes

File: python/foglamp/plugins/south/http_south/http_south.py (hunk 7 above)
Comment (amarendra-dianomic) [reply python/foglamp/plugins/south/http_south/http_south.py]: yes

File: python/foglamp/plugins/south/http_south/http_south.py (hunk 9 above)
Comment (amarendra-dianomic) [reply python/foglamp/plugins/south/http_south/http_south.py]: explained above.

File: python/foglamp/plugins/south/http_south/http_south.py (hunk 8 above)
Comment (amarendra-dianomic) [reply python/foglamp/plugins/south/http_south/http_south.py]: explained above

File: python/foglamp/plugins/south/http_south/http_south.py (hunk 10 above)
Comment (amarendra-dianomic) [reply python/foglamp/plugins/south/http_south/http_south.py]: To maintain uniformity amongst all South plugin shutdown methods.

File: python/foglamp/plugins/south/http_south/http_south.py (hunk 11 above)
Comment (amarendra-dianomic) [reply python/foglamp/plugins/south/http_south/http_south.py]: Just to make the example shorter and clearer.

File: python/foglamp/services/south/server.py (hunk 13 above)
Comment (amarendra-dianomic) [reply python/foglamp/services/south/server.py]: for pexpect.exceptions.TIMEOUT handling

**Comment by amarendra-dianomic** [comment]
@praveen-garg all changes done per feedback. Please check.

File: python/foglamp/plugins/utils.py [hunk 23]
```diff
@@ -0,0 +1,28 @@
+# -*- coding: utf-8 -*-
```
Comment (MarkRiddoch) [inline python/foglamp/plugins/utils.py]: This file should be in python/foglamp/plugins/common/utils.py

**Comment by praveen-garg** [comment]
Looks fine to me, @Singhal-Vaibhav  can we please test and close the corresponding tickets?!

- Make sure we test,  restart is conditional for port, address only

**Pull request merged**